### ISSUE TITLE
INVITE API creates case sensitive e-mail accounts

### ISSUE THREAD

**Issue by tiblu** [citizenos/citizenos-api#234]
**What is the problem?**

If a User is invited to Topic/Group, the e-mail is case-sensitive so for example `test@citizenos.com` and `Test@citizenos.com` are considered different accounts.
When there is a capitalization difference, a new account is created rather than using existing.
When you invite User with `Test@citizenos.com`, they cannot see the Topic/Group when logging in with `test@citizenos.com`.

**Why is this a problem?**

When you invite User with `Test@citizenos.com`, they cannot see the Topic/Group when logging in with `test@citizenos.com`.

**Possible solution.**

- [x] Find all API-s that create User (`User.create`, `User.findOrCreate`...) and write a test to verify the bugs existence. Then FIX it.
    - [x] Topic invites failing testcase 
    - [x] Group invites  failing testcase
- [x] Make sure `User` model ALWAYS makes e-mail case-insensitive (lowercase).
    - [x] Probably can use the same pattern we used to hash passwords.
- [x] Make sure e-mails are sent out using case-insensitive address. Email system SHOULD be insensitive by default, but you never know what's downstream
- [ ] Migration - there are about 20 accounts that seem to have duplicates due to case-sensitivity. Decide what to do. I THINK it would be fair to migrate all content under one account, BUT might be fair to just DELETE/obfuscate duplicate email accounts IF the accout has no content. WE CANNOT skip migration as login is now case-insensitivie and we have no idea which account is picked up when User logs in. Query to find out which accounts have issues: `SELECT LOWER(email), COUNT(email) FROM "Users" WHERE email IS NOT NULL GROUP BY 1 HAVING COUNT(email) > 1 ORDER BY 2 DESC;`

**NOTES**

* Originally discovered by @BeccaMelhuish and discussed in Slack on the 3rd of February 2022.

**Comment by anettlinno** [comment]
Triage 60. Important update. Est. dev. time 8 hours. Sending to development.

**Comment by tiblu** [comment]
## Migration

The migration logic will be following:

* Find all Users with duplicate e-mails with different capitalization.
* Find which account of the duplicates has most Activities.
* Mark all other accounts as deleted. (`deletedAt`).
* Record account delete activity with the reference to the migration.

### Why we do not migrate data to 1 account from the duplicates?

1. Most likely the most active account is the one that sees use.
2. It is likely that if User was invited, they did not see the invitation anyway due to the capitalization difference in e-mail and was re-invited.
3. We cannot alter vote history.
4. Resolving conflicts where User has access to a resource with different accounts is going to be expensive.

TLDR: Migration const vs benefit ratio makes it unfeasible.

**Comment by tiblu** [comment]
### Release 05.03.2022 01:27 - Migration

https://github.com/citizenos/citizenos-api/blob/37decc3c2747ae53197115e8a566759a3180eaf8/db/migrations/20220228174313-duplicate-email-users-issue-234.js

* DB backup - **b2299**
* Unique Users migrated: 36
* Total Users marked as deleted: 37

#### How can I spot a migrated User?

* Migrated Users - `SELECT id, email FROM "Users" WHERE "email" SIMILAR TO '%\_\_migrated\_\d{1,}';`
* Activities - `SELECT COUNT(1) FROM "Activities" WHERE data->>'__migrationId' = '20220228174313-duplicate-email-users-issue-234.js';`
